src/pybind/mgr/alerts/module.py

```diff
@@ -0,0 +1,243 @@
+                                  new_status)
+                self.log.debug('diff %s' % diff)
+                if diff:
+                    self._send_alert(new_status, diff)
+                last_status = new_status
+
+            # Use mgr_tick_period (default: 2) here just to illustrate
+            # consuming native ceph options.  Any real background work
+            # would presumably have some more appropriate frequency.
+            self.log.debug('Sleeping for %d seconds', self.interval)
+            ret = self.event.wait(self.interval)
+            self.event.clear()
+
+    def shutdown(self):
+        """
+        This method is called by the mgr when the module needs to shut
+        down (i.e., when the serve() function needs to exit).
+        """
+        self.log.info('Stopping')
+        self.run = False
+        self.event.set()
+
+    # SMTP
+    def _smtp_format_alert(self, code, alert):
+        r = '[{sev}] {code}: {summary}\n'.format(
+            code=code,
+            sev=alert['severity'].split('_')[1],
+            summary=alert['summary']['message'])
+        for detail in alert['detail']:
+            r += '        {message}\n'.format(
+                message=detail['message'])
+        return r
+
+    def _send_alert_smtp(self, status, diff):
+        # message
+        self.log.debug('_send_alert_smtp')
+        message = ('From: {from_name} <{sender}>\n'
+                   'Subject: {status}\n'
+                   'To: {target}\n'
+                   '\n'
+                   '{status}\n'.format(
+                       sender=self.smtp_sender,
+                       from_name=self.smtp_from_name,
+                       status=status['status'],
+                       target=self.smtp_destination))
+
+        if 'new' in diff:
+            message += ('\n--- New ---\n')
+            for code, alert in diff['new'].items():
+                message += self._smtp_format_alert(code, alert)
+        if 'updated' in diff:
+            message += ('\n--- Updated ---\n')
+            for code, alert in diff['updated'].items():
+                message += self._smtp_format_alert(code, alert)
+        if 'cleared' in diff:
+            message += ('\n--- Cleared ---\n')
+            for code, alert in diff['cleared'].items():
+                message += self._smtp_format_alert(code, alert)
+
+        message += ('\n\n=== Full health status ===\n')
+        for code, alert in status['checks'].items():
+            message += self._smtp_format_alert(code, alert)
+            
+        self.log.debug('message: %s' % message)
+
+        # send
+        if self.smtp_ssl:
+            server = smtplib.SMTP_SSL(self.smtp_host, self.smtp_port)
+        else:
+            server = smtplib.SMTP(self.smtp_host, self.smtp_port)
+        if self.smtp_password:
+            server.login(self.smtp_user, self.smtp_password)
+        server.sendmail(self.smtp_sender, self.smtp_destination, message)
+        server.quit()
```

Comment:
The smtplib methods (including the constructor) will raise various exceptions in case of connection refused, host not found and SMTP errors. Maybe wrap the whole lot in a try/except, which just catches an `Exception`, then logs `str(ex)` as an error?

Reply:
Yeah... getting this to work involved a bit of trial and error.

The bigger problem IMO is that when you have code that throws an exception in the command handling path, you see the traceback in stderr on the CLI, and can debug.  But if there is an exception in serve() code, the mgr long *only* includes the exception name/string itself (e.g., "KeyError 1," or something), but not the backtrace.  Any idea how we can fix that?

For this, I think I should catch any smtplib exception and raise a health alert.  I'll give it a go...

Reply:
> the mgr long _only_ includes the exception name/string itself (e.g., "KeyError 1," or something), but not the backtrace

That sounds like a bug. `handle_pyerror()` in PyModule.cc is meant to turn a python Exception into a string (including the backtrace). This is called from various places inside mgr to dump exceptions to the error log.